include/boost/geometry/algorithms/detail/overlay/get_clusters.hpp

```diff
@@ -0,0 +1,302 @@
+    }
+
+};
+
+#ifdef BOOST_GEOMETRY_USE_RESCALING_IN_GET_CLUSTERS
+template <>
+struct sweep_equal_policy<boost::long_long_type>
+{
+    template <typename P>
+    static inline bool equals(P const& p1, P const& p2)
+    {
+        // Neighbouring cells in the "integer grid" are considered as equal
+        return std::abs(geometry::get<0>(p1) - geometry::get<0>(p2)) <= 1
+            && std::abs(geometry::get<1>(p1) - geometry::get<1>(p2)) <= 1;
+    }
+
+    static inline bool exceeds(boost::long_long_type value)
+    {
+        return value > 1;
+    }
+
+};
+#endif
+
+template <typename Point>
+struct turn_with_point
+{
+    std::size_t turn_index;
+    Point pnt;
+};
+
+template <typename Cluster, typename Point>
+struct cluster_with_point
+{
+    Cluster cluster;
+    Point pnt;
+};
+
+// Use a sweep algorithm to detect clusters
+template
+<
+    typename Turns,
+    typename Clusters,
+    typename RobustPolicy
+>
+inline void get_clusters(Turns& turns, Clusters& clusters, RobustPolicy const& robust_policy)
+{
+    using turn_type = typename boost::range_value<Turns>::type;
+    using cluster_type = typename Clusters::mapped_type;
+
+#ifdef BOOST_GEOMETRY_USE_RESCALING_IN_GET_CLUSTERS
+    // For now still use robust points for rescaled, otherwise points do not match
+    using point_type = typename geometry::robust_point_type
+    <
+        typename turn_type::point_type,
+        RobustPolicy
+    >::type;
+#else
+    using point_type = typename turn_type::point_type;
+#endif
+
+    sweep_equal_policy<typename coordinate_type<point_type>::type> equal_policy;
+
+    std::vector<turn_with_point<point_type>> points;
+    std::size_t turn_index = 0;
+    for (auto const& turn : turns)
+    {
+        if (! turn.discarded)
+        {
+#ifdef BOOST_GEOMETRY_USE_RESCALING_IN_GET_CLUSTERS
+            point_type pnt;
+            geometry::recalculate(pnt, turn.point, robust_policy);
+            points.push_back({turn_index, pnt});
+#else
+            points.push_back({turn_index, turn.point});
+#endif
+        }
+        turn_index++;
+    }
+
+    // Sort the points from top to bottom
+    std::sort(points.begin(), points.end(), [](auto const& e1, auto const& e2)
+    {
+       return geometry::get<1>(e1.pnt) > geometry::get<1>(e2.pnt);
+    });
+
+    // The output vector will be sorted from bottom too
+    std::vector<cluster_with_point<cluster_type, point_type>> clustered_points;
+
+    // Compare points with each other. Performance is O(n log(n)) because of the sorting.
+    int cnt = 0;
+    for (auto it1 = points.begin(); it1 != points.end(); ++it1)
+    {
+        // Inner loop, walks until too far away in y-direction
+        for (auto it2 = it1 + 1; it2 != points.end(); ++it2)
+        {
+            auto const d = geometry::get<1>(it1->pnt) - geometry::get<1>(it2->pnt);
+            if (equal_policy.exceeds(d))
+            {
+                // Too far away in y-direction to be considered as equal
+                break;
```

Comment:
Why is y direction special?

Reply:
Yes, it's an arbitrary choice.

Clusters are locations with multiple turns. They are crucial to the traversal algorithm. It should work for either cartesian or non cartesian coordinates.

I don't understand that x-coordinate would be better, in my opinion y-direction is better just because of those antimeridian points.
It's all turn-based, so it's point-based. There are no segments or polygons involved (in the clusters). Of course turns are calculated from lines or polygons intersections, and they may span anything.

Polygons surrounding poles or crossing the antimeridian are not problematic, unless there would be a cluster at (90) or (-90) (one point). I think that's a theoretical problem, it won't happen in practice (unless you make a testcase for it). The chance for a cluster on the -180 meridian is low too, but higher because it's a globe crossing line.

Reply:
> Clusters are locations with multiple turns.

So these are points that may be close to each other. These points are either endpoints or intersection points. And the purpose of this scanning is to find them and group close ones into clusters. Correct?

If that's the case then I think indeed this particular algorithm should work in any CS. I guess the coordinates might be normalized but technically valid range is in (-180 -90, 180 90) so ok.

----

I'll explain what I was thinking about regarding vertices, segments, etc. Consider this:

![path10](https://user-images.githubusercontent.com/1226951/112411043-87178b80-8d1c-11eb-81f4-150419cc08dd.png)

In spherical/geographic CS, points of a segment can lie "above" its endpoints and any detected intersection point on this segment unless the intersection point is at the vertex or vertex of a geodesic is not between the endpoints. So in a hypothetical scenario of looking for segments/rings/etc. by scanning for endpoints or intersection points it is not possible to detect the segment correctly when the algorithm is scanning the latitudes.

But since your goal (AFAIU) is only to group close points together, this is not a problem I think.

Reply:
> And the purpose of this scanning is to find them and group close ones into clusters. Correct?

Correct

> I'll explain what I was thinking about regarding vertices, segments, etc.

Thanks! It's good to keep in mind if this algorithm (the sweeping part) is used more often. And yes, I then see what you mean and going in x-direction will be more convenient.

Reply:
Ok, so the only problem is the normalization/winding because any point with latitude close to -90 or 90 should be considered equal to any other regardless of the longitude. And points with londitude -180 should be considered equal to points with longitude 180. The normalization functions we have would AFAIR set the longitudes to 0 for points at poles and longitude to 180 if points are -180. All of this is done WRT machine epsilon, this is done for points close to poles or antimeridian. So if your epsilon is greater than the machine epsilon then you'd have to normalize coordinates differently or take the CS into consideration by manually checking the closeness in corresponding way.

Reply:
I started wondering about one more thing. Since this is only for the checks if points are close to each other. Have you considered to do what we usually do when checking points of equality (which is actually closeness WRT machine epsilon), i.e. simply using `std::set` with a custom `less` making sure that close points are considered equal?

Reply:
It's a good idea. I'll check it on Wednesday together with possible epsilon changes. Even with the sorting (which indeed should stay), I maybe can still use the set for the clusters.

Reply:
I have had a look. I'm not finished yet but I doubt if it is more convenient. 

Can you point me to the code where we actually really do this? Because with a custom less operator in two dimensions, points approximately considered equal, it's not that trivial. I've some testcode which (still) results in slightly different results and is more complex than what I had written originally.

However, a unit test is added already and that is useful anyway. Plus code is split better. (see last commit)

Reply:
Ok, not `std::set` but `std::binary_search` but it's the same story: https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp#L80-L83
And I agree that this is technically incorrect because for 3 points: with coord1 = 0, coord2 = eps and coord3 = 2eps. The first one is equal to the second, the second to the third but the third is not equal to the first one. Which is inconsistent and I guess might cause assertion failure in some STL implementations.

Reply:
OK, thanks. Then I keep it as it was.

Thanks for the review comments and the approval.